### PR TITLE
feat: added `idProvider` to model constructor

### PR DESCRIPTION
**Pull request by MacMcIrish**

### CODE REVIEW

File: src/modules/model.js [hunk 1]
```diff
@@ -18,7 +18,8 @@ class Model {
       ItemNotFound = DefaultItemNotFound,
       ItemExists = DefaultItemExists
     } = {},
-    callback = () => {}
+    callback = () => {},
+    idProvider = null
```
Comment (simlu) [inline src/modules/model.js]: add assert

File: src/modules/errors.js [hunk 2]
```diff
@@ -32,3 +32,24 @@ class ConditionNotImplemented extends Error {
...
 module.exports.ConditionNotImplemented = ConditionNotImplemented;
+
+class CannotUpdatePrimaryKeys extends Error {
+  constructor() {
+    super('Cannot update "primaryKeys" attributes.');
+  }
+}
+module.exports.CannotUpdatePrimaryKeys = CannotUpdatePrimaryKeys;
+
+class MustProvideIdOrPrimaryKeys extends Error {
```
Comment (simlu) [inline src/modules/errors.js]:
```suggestion
class MustProvideIdXorPrimaryKeys extends Error {
```

File: README.md [hunk 3]
```diff
@@ -85,7 +87,7 @@ modelName.create({ id, data, fields });
 
 _Params_
 
-* `id` string: Id of model to create, must be unique
+* `id` string: Id of model to create, must be unique. _Optional if primaryKeys configured_.
```
Comment (simlu) [inline README.md]: fix

File: src/modules/model.js [hunk 4]
```diff
@@ -66,6 +74,21 @@ class Model {
     });
   }
 
+  // eslint-disable-next-line no-underscore-dangle
+  _generateId(data, providedId) {
+    if ((providedId === null) === (this.primaryKeys === null)) {
```
Comment (simlu) [inline src/modules/model.js]:
```suggestion
    if ((typeof providedId !== 'string') === (!Array.isArray(this.primaryKeys))) {
```

plus add more asserts?

File: src/modules/model.js [hunk 5]
```diff
@@ -66,6 +74,21 @@ class Model {
...
+      throw new MustProvideIdXorPrimaryKeys();
+    }
+    if (this.primaryKeys !== null && this.primaryKeys.some(k => data[k] === undefined)) {
```
Comment (simlu) [inline src/modules/model.js]:
```suggestion
    if (Array.isArray(this.primaryKey) && this.primaryKeys.some(k => data[k] === undefined)) {
```

File: src/modules/model.js [hunk 6]
```diff
@@ -66,6 +74,21 @@ class Model {
...
+      throw new IncompletePrimaryKey();
+    }
+    return providedId !== null
```
Comment (simlu) [inline src/modules/model.js]:
```suggestion
    return typeof providedId === 'string'
```

File: src/modules/model.js [hunk 7]
```diff
@@ -66,6 +74,21 @@ class Model {
...
+      ? providedId
+      : objectHash(this.primaryKeys
+        .sort()
```
Comment (simlu) [inline src/modules/model.js]: I don't think objectHash differentiates object key order?

File: src/modules/model.js [hunk 8]
```diff
@@ -113,6 +142,9 @@ class Model {
   }) {
     // eslint-disable-next-line no-underscore-dangle
     this._before();
+    if (this.primaryKeys !== null && this.primaryKeys.some(k => data[k] !== undefined)) {
```
Comment (simlu) [inline src/modules/model.js]: as above

File: src/modules/model.js (hunk 7 above)
Comment (MacMcIrish) [reply src/modules/model.js]: Ah, you're right. They sort by default, I only saw the option

File: README.md [hunk 9]
```diff
@@ -85,7 +87,7 @@ modelName.create({ id, data, fields });
 
 _Params_
 
-* `id` string: Id of model to create, must be unique
+* `id` string: Id of model to create, must be unique. _Must provide id or configure primaryKeys but not both_.
```
Comment (simlu) [inline README.md]:
```suggestion
* `id` string: Id of model to create, must be unique. _Required iff primaryKeys no provided_.
```

File: src/modules/model.js [hunk 10]
```diff
@@ -66,6 +75,22 @@ class Model {
     });
   }
 
+  // eslint-disable-next-line no-underscore-dangle
+  _generateId(data, providedId) {
+    assert(typeof data === 'object' && !Array.isArray(data));
+    assert(providedId === null || typeof providedId === 'string');
+    assert(this.primaryKeys === null || Array.isArray(this.primaryKeys));
+    if ((providedId === null) === (!Array.isArray(this.primaryKeys))) {
```
Comment (simlu) [inline src/modules/model.js]: ?

File: src/modules/model.js [hunk 11]
```diff
@@ -66,6 +75,22 @@ class Model {
     });
   }
 
+  // eslint-disable-next-line no-underscore-dangle
+  _generateId(data, providedId) {
+    assert(typeof data === 'object' && !Array.isArray(data));
```
Comment (simlu) [inline src/modules/model.js]: fix

File: src/modules/model.js [hunk 12]
```diff
@@ -135,6 +172,25 @@ class Model {
...
+    fields,
+    conditions = []
+  }) {
+    // eslint-disable-next-line no-underscore-dangle
+    this._before();
+    // eslint-disable-next-line no-underscore-dangle
+    const id = this._generateId(data, providedId);
+    await this.mapper.put(new this.MapperClass({ ...data, id }), Object.assign(
+      {},
+      conditions.length === 0 ? {} : { condition: { type: 'And', conditions: [...conditions] } }
```
Comment (simlu) [inline src/modules/model.js]: fix

File: test/modules/model.spec.js [hunk 13]
```diff
@@ -63,6 +66,16 @@ describe('Dynamo Sdk Tests', () => {
       ItemNotFound: CustomItemNotFound
     }
   });
+  const autoIdModel = new DynamoModel({
```
Comment (simlu) [inline test/modules/model.spec.js]: rename?

File: src/modules/model.js [hunk 14]
```diff
@@ -66,6 +75,22 @@ class Model {
     });
   }
 
+  // eslint-disable-next-line no-underscore-dangle
+  _generateId(data, providedId) {
+    assert(data instanceof Object && !Array.isArray(data), data);
+    assert(providedId === null || typeof providedId === 'string');
+    assert(this.primaryKeys === null || Array.isArray(this.primaryKeys));
```
Comment (simlu) [inline src/modules/model.js]:
```suggestion
```

**Pull request merged**